pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -158,22 +193,143 @@ func (n *nodeLocalDNS) Deploy(ctx context.Context) error {
+	if err := managedresources.DeleteForShoot(ctx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
+
+	if v1beta1helper.IsKubeProxyIPVSMode(n.values.KubeProxyConfig) {
+		return nil
+	}
+
+	// Wait until the managed resource is successfully deleted and go through node list again to add cleanup label
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
+
+	nodeList = &metav1.PartialObjectMetadataList{}
+	nodeList.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("NodeList"))
+	if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabelsSelector{Selector: labels.NewSelector().Add(utils.MustNewRequirement(labelKeyCleanupRequired, selection.DoesNotExist))}); err != nil {
+		return fmt.Errorf("failed to list all nodes without %s label: %w", labelKeyCleanupRequired, err)
+	}
+
+	var taskFns []flow.TaskFn
+	if managedResourceExists {
+		for _, node := range nodeList.Items {
+			taskFns = append(taskFns, func(ctx context.Context) error {
+				patch := client.MergeFrom(node.DeepCopy())
+				metav1.SetMetaDataLabel(&node.ObjectMeta, labelKeyCleanupRequired, "true")
+				if err := n.values.ShootClient.Patch(ctx, &node, patch); err != nil {
+					return fmt.Errorf("failed to add cleanup label to recently joined node %s: %w", node.Name, err)
+				}
+				return nil
+			})
+		}
+	}
+
+	if err := flow.Parallel(taskFns...)(ctx); err != nil {
+		return fmt.Errorf("failed to remove cleanup label from nodes: %w", err)
+	}
+
+	if err := n.createCleanupConfigMap(ctx); err != nil {
+		return fmt.Errorf("failed to create cleanup ConfigMap: %w", err)
+	}
+
+	if err := n.createCleanupDaemonSet(ctx); err != nil {
+		return fmt.Errorf("failed to create cleanup DaemonSet: %w", err)
+	}
+
+	if err := n.waitForDaemonSetCompletion(ctx, metav1.NamespaceSystem, labelValueAndCleanupName); err != nil {
+		return fmt.Errorf("cleanup DaemonSet %s failed: %w", labelValueAndCleanupName, err)
+	}
+
+	n.values.Log.Info("Cleanup DaemonSet for node-local-dns completed")
+	cleanupDaemonSet := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      labelValueAndCleanupName,
+			Namespace: metav1.NamespaceSystem,
+		},
+	}
+	if err := n.values.ShootClient.Delete(ctx, cleanupDaemonSet); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("failed to delete cleanup DaemonSet %s: %w", labelValueAndCleanupName, err)
+	}
+
+	taskFns = []flow.TaskFn{}
+	nodeList = &metav1.PartialObjectMetadataList{}
+	nodeList.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("NodeList"))
+	if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabels{
+		labelKeyCleanupRequired: "true",
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes for cleanup: %w", err)
+	}
+	// Remove the cleanup-required label from all nodes in this pool in parallel using flow
```

Comment:
```suggestion

```

You don't have such comment before - it's also not needed, the code is well-readable now